bindings/matrix-sdk-ffi/src/timeline/mod.rs

```diff
@@ -521,6 +526,51 @@ impl Timeline {
         Ok(())
     }
 
+    /// Sends a user's location as a beacon based on their last beacon_info
+    /// event.
+    ///
+    /// Retrieves the last beacon_info from the room state and sends a beacon
+    /// with the geo_uri. Since only one active beacon_info per user per
+    /// room is allowed, we can grab the currently live beacon_info event
+    /// from the room state.
+    ///
+    /// TODO: Does the logic belong in self.inner.room() or here?
+    pub async fn send_user_location_beacon(
+        self: Arc<Self>,
+        geo_uri: String,
+    ) -> Result<(), ClientError> {
```

Comment:
There's way too much logic in this function; it should not live at the FFI layer, otherwise it's ~impossible to test.